Generate extra enum method helpers for Ruby

jbolinger opened the pull request:
This is an example of a potential solution to #4598. It causes the generator to create an extra method for types with enum fields in the following form:

```ruby
class MyMessage
  def type_const
    Google::Example::Resource::Type.const_get(type)
  end
end
```

`type` is the field name of an enum type in the `MyMessage` proto message in this example.

I'm not sure if there's a better way to add something like this via the DSL (I couldn't find any docs).

TeBoring commented:
Tests still failing

jbolinger commented:
Opps, I missed those tests but I've updated them now.

The generated code could be simplified a bit here, but I wanted to get your feedback on the approach first.

TeBoring commented:
Could you add a test to use the new method?

jbolinger commented:
Certainly! I've added one.

jbolinger commented:
@TeBoring @haberman what do you think about the way these additional helper methods are generated? I was thinking of taking the same approach for the `*_as_value` methods that we've been discussing.

TeBoring commented:
The approach looks good to me. I am just not sure whether the name is idiomatic to ruby users.
I also want @haberman to take a look to make sure we want this API.

blowmage commented on src/google/protobuf/compiler/ruby/ruby_generated_code_pb.rb:
```diff
@@ -70,6 +70,17 @@ module B
       TestMessage = Google::Protobuf::DescriptorPool.generated_pool.lookup("A.B.C.TestMessage").msgclass
       TestMessage::NestedMessage = Google::Protobuf::DescriptorPool.generated_pool.lookup("A.B.C.TestMessage.NestedMessage").msgclass
+      class TestMessage
+        def optional_enum_const
```
I am not a fan of adding these methods by altering the generated Ruby code. I would rather see these new methods added to the object that is returned when calling `lookup("A.B.C.TestMessage").msgclass`. That way, Ruby files that were previously generated will get these new methods by upgrading to the latest google-protobuf gem, without having to regenerate the the Ruby files.

jbolinger commented:
I updated the implementation per @blowmage's suggestion. The method functionality remains the same but the generated code is no longer changed. I also added some more tests for various kinds of field types.

blowmage commented on ruby/tests/common_tests.rb:
```diff
@@ -706,6 +706,50 @@ def test_enum_lookup
+    assert_equal :B, m.optional_enum
+    assert_equal 2, m.optional_enum_const
+    assert_equal [:A, :C], m.repeated_enum
+    assert_equal [1, 3], m.repeated_enum_const
```
Is this the same as the following?

```ruby
assert_equal [proto_module::TestEnum::A, proto_module::TestEnum::C], m.repeated_enum_const
```

If so, would it be possible to show the constants as well as the constant's values?

jbolinger replied:
Yes, they are the same and it would be clearer to show that. I'll add both to the tests.

blowmage replied:
:heart:

TeBoring commented on ruby/ext/google/protobuf_c/message.c:
```diff
@@ -231,13 +255,13 @@ VALUE Message_method_missing(int argc, VALUE* argv, VALUE _self) {
       // METHOD_ACCESSOR
       return oneof_field == NULL ? Qnil :
-	ID2SYM(rb_intern(upb_fielddef_name(oneof_field)));
+       ID2SYM(rb_intern(upb_fielddef_name(oneof_field)));
```
Is the indent correct?

jbolinger replied:
fixed.

TeBoring commented on ruby/ext/google/protobuf_c/message.c:
```diff
@@ -153,9 +154,32 @@ static int extract_method_call(VALUE method_name, MessageHeader* self,
+			                                   &test_f, &test_o);
+
+  // Look for enum accessor of the form <enum_name>_const
+  if (!has_field && accessor_type == METHOD_GETTER && name_len > 6) {
```
Is this checking too loose?
What if another field's name is ```foo_XXXXX```?

TeBoring replied:
Please also test this case.

jbolinger replied:
I think it's okay because `METHOD_UNKNOWN` will still be returned unless the `enum_name` is found in the innermost check. I added that as a test case - is that what you were concerned about?

TeBoring replied:
I think it matters.
Suppose foo is the enum field, if you call m.foo_abcde, you don't return 
METHOD_UNKNOWN.

jbolinger replied:
Oh, I see. Good catch. I've added an additional check for the `_const` suffix.

TeBoring commented on ruby/tests/common_tests.rb:
```diff
@@ -722,6 +722,16 @@ def test_enum_const_get_helpers
     assert_equal proto_module::TestEnum::C, m.optional_enum_const
     assert_equal proto_module::TestEnum.const_get(:C), m.optional_enum_const
+
+    m = proto_module::TestMessage2.new({foo: 2})
```
Foo needs to be an enum field.

jbolinger replied:
Added a new test that uses an enum field.

TeBoring commented:
please sync

jbolinger commented:
@TeBoring sync'd. Thanks for the help!
